conda/core/package_cache_data.py

```diff
@@ -729,10 +824,67 @@ def __eq__(self, other):
         return hash(self) == hash(other)
 
 
+# called from ProgressiveFetchExtract.execute()
+def do_cache_action(prec, cache_axn, progress_bar, download_total=1.0):
+    # dummy action to simplify code
+    if not cache_axn:
+        return prec
+    cache_axn.verify()
+
+    if not cache_axn.url.startswith("file:/"):
+
+        def progress_update_cache_axn(pct_completed):
+            progress_bar.update_to(pct_completed * download_total)
+
+    else:
+        download_total = 0
+        progress_update_cache_axn = None
+
+    cache_axn.execute(progress_update_cache_axn)
+    return prec
+
+
+def do_extract_action(prec, extract_axn, progress_bar):
+    # dummy action
+    if not extract_axn:
+        return prec
+    extract_axn.verify()
+    # currently unable to do updates on extract;
+    # likely too fast to bother
+    extract_axn.execute(None)
+    progress_bar.update_to(1.0)
+    return prec
+
+
+def do_cleanup(progress_bar, *actions):
+    for action in actions:
+        if action:
+            action.cleanup()
+
+
+def do_reverse(progress_bar, *actions):
+    for action in actions:
+        if action:
+            action.reverse()
+
+
+def done_callback(future, actions, progress_bar: ProgressBar, exceptions: list, finish=False):
```

Comment:
```suggestion
def done_callback(future: Future, actions: tuple[Union[CacheUrlAction, ExtractPackageAction], ...], progress_bar: ProgressBar, exceptions: list[Exception], finish: bool = False):
```

Might as well go all the way if you're going to use type hints 😅 . If you decide to incorporate this change, be sure to add the following to the top of the file:

```python
from __future__ import annotations

from typing import Union
```

Reply:
Also, this function would benefit from a doc string too 🙏